bed/enrichment.go

```diff
@@ -0,0 +1,125 @@
+package bed
+
+import (
+	"github.com/vertgenlab/gonomics/numbers"
+	"math"
+	"strings"
+)
+
+//subtractFromBedCoord is a helper function of overlapProbability that subtracts the values subStart and subEnd from entries in a slice of Bed b.
+//bClone is a clone of b used to save on memory allocation.
+func subtractFromBedCoord(b []*Bed, subStart int, subEnd int, bClone []*Bed) {
+	var prevEnd int = 0
+	var prevChrom string = ""
+
+	for i := range b {
+		if strings.Compare(prevChrom, "") == 0 || strings.Compare(prevChrom, bClone[i].Chrom) != 0 {
+			prevChrom = bClone[i].Chrom
+			prevEnd = 0
+		}
+		bClone[i].Chrom = b[i].Chrom
+		bClone[i].ChromStart = numbers.Max(prevEnd, b[i].ChromStart-subStart)
+		bClone[i].ChromEnd = numbers.Max(b[i].ChromStart, b[i].ChromEnd-subEnd)
+		prevEnd = bClone[i].ChromEnd
+	}
+}
+
+//overlapProbability calculates the probability that an element of len 'length' overlaps a set of genomic 'elements' in a genome represented by 'noGapRegions'.
+//tempElements and tempNoGap represent cloned slices of elements and noGapRegions which are used for memory optimization.
+func overlapProbability(elements []*Bed, tempElements []*Bed, length int, noGapRegions []*Bed, tempNoGap []*Bed) float64 {
+	subtractFromBedCoord(elements, length-1, 0, tempElements)
+	subtractFromBedCoord(noGapRegions, 0, length-1, tempNoGap)
+	return float64(OverlapLengthSum(tempElements, tempNoGap)) / float64(TotalSize(tempNoGap))
+}
+
+func ElementOverlapProbabilities(elements1 []*Bed, elements2 []*Bed, noGapRegions []*Bed) []float64 {
+	var tempElements1 []*Bed = make([]*Bed, len(elements1))
+	var tempElements2 []*Bed = make([]*Bed, len(elements2))
+	var tempNoGap []*Bed = make([]*Bed, len(noGapRegions))
+	var answer []float64 = make([]float64, len(elements2))
+	copy(tempElements1, elements1)
```

Comment:
I notice that these copies need to be not only on the pointers, but on the data/structs themselves.  The way it is coded now, we are modifying the original data.